Add framework docs (#1775)
hashicorp/terraform-website: content/source/docs/plugin/framework/accessing-values.html.md

---
layout: "extend"
page_title: "Plugin Development - Framework: Access State, Config, and Plan"
description: |-
  How to read values from state, config, and plan in the next-generation
  Terraform provider development framework.
---

# Access State, Config, and Plan

There are various points at which the provider needs access to the data from

> the practitioner's configuration, Terraform's state, or the plan that was
> generated. The same patterns are used for accessing this data, regardless of

Reviewer: ```suggestion
the practitioner's configuration, Terraform's state, or generated plan.
The same patterns are used for accessing this data, regardless of
```

its source.

The data is usually stored in a request object:

```go
func (m myResource) Create(ctx context.Context, req tfsdk.CreateResourceRequest, resp *tfsdk.CreateResourceResponse)
```

> In this example, `req` holds the configuration and plan; there is no state
> value, as the resource doesn't exist in state yet (that's why it's being
> created).

Reviewer 2: ```suggestion
In this example, `req` holds the configuration and plan, and there is no state
value because the resource does not yet exist in state.
```

code that accesses values, but requires defining a type to contain the values.

```go
type resourceData struct {
	Name types.String `tfsdk:"name"`
	Age types.Number `tfsdk:"age"`
	Registered types.Bool `tfsdk:"registered"`
	Pets types.List `tfsdk:"pets"`
	Tags types.Map `tfsdk:"tags"`
	Address types.Object `tfsdk:"address"`
}

func (m myResource) Create(ctx context.Context, req tfsdk.CreateResourceRequest, resp *tfsdk.CreateResourceResponse) {
	var plan resourceData
	err := req.Plan.Get(ctx, &plan)
	if err != nil {
		// TODO: handle error
	}
	// values can now be accessed like plan.Name.Value
	// check if things are null with plan.Name.Null
	// check if things are unknown with plan.Name.Unknown
}
```

The configuration, plan, and state data is represented as an object, and
accessed like an object. See the [conversion rules](#conversion-rules) for an
explanation on how objects can be converted into Go types.

Using the [`attr.Value` implementations](/docs/plugin/framework/types.html) can
surface complexity that is unnecessary, however. In a create function,
non-computed values are guaranteed not to be unknown, for example, so there's

> no need to worry about them. Likewise, a required value will never be null, and

Reviewer: can we rewrite this sentence so it doesn't use a double negative "guaranteed not to be unknown"

> Using the [`attr.Value` implementations](/docs/plugin/framework/types.html) can
> surface complexity that is unnecessary, however. In a create function,
> non-computed values are guaranteed not to be unknown, for example, so there's
> no need to worry about them. Likewise, a required value will never be null, and
> that complexity is also unnecessary.

Reviewer: ```suggestion
However, using the [`attr.Value` implementations](/docs/plugin/framework/types.html) 
can surface unnecessary complexity. For example, in a create function,
non-computed values are guaranteed to be defined. Likewise, a required value 
will never be null.
```

it, like `type MyString string`) as long as the string value isn't null or
unknown.

### Number

Numbers can be automatically converted to the following numeric types (or any
aliases of them, like `type MyNumber int`) as long as the number value isn't
null or unknown:

* `int`, `int8`, `int16`, `int32`, `int64`
* `uint`, `uint8`, `uint16`, `uint32`, `uint64`
* `float32`, `float64`
* [`*big.Int`](https://golang.org/pkg/math/big#Int), [`*big.Float`](https://golang.org/pkg/math/big#Float)

If the value of the number can't be stored in the numeric type supplied, either
because of an overflow or other loss of precision, an error will be returned.

### Boolean

Booleans can be automatically converted to Go's `bool` type (or any aliases of
it, like `type MyBoolean bool`) as long as the boolean value isn't null or
unknown.

### List

Lists can be automatically converted to any Go slice type (or alias of a Go
slice type, like `type MyList []string`), with the elements either being
`attr.Value` implementations or being converted according to these rules. Go
slice types are considered capable of handling null values; the slice will be

> set to nil. Unknown list values will still cause an error to be returned.

Author: The `Get` method.

> set to nil. Unknown list values will still cause an error to be returned.

Reviewer 2: thank you!

### Map

Maps can be automatically converted to any Go map type with string keys (or any
alias of a Go map type with string keys, like `type MyMap map[string]int`),
with the elements either being `attr.Value` implementations or being converted
according to these rules. Go map types are considered capable of handling null

> values; the map will be set to nil. Unknown map values will still cause an

Author: Same answer, the `Get` method. Want me to update it to fix, or would you like to do the honors? :D

> values; the map will be set to nil. Unknown map values will still cause an

Reviewer 2: heh I have left a suggested revision - thank you!!!

it, like `type MyString string`) as long as the string value is not null or
unknown.

### Number

Numbers can be automatically converted to the following numeric types (or any
aliases of them, like `type MyNumber int`) as long as the number value is not
null or unknown:

* `int`, `int8`, `int16`, `int32`, `int64`
* `uint`, `uint8`, `uint16`, `uint32`, `uint64`
* `float32`, `float64`
* [`*big.Int`](https://golang.org/pkg/math/big#Int), [`*big.Float`](https://golang.org/pkg/math/big#Float)

An error will be returned if the value of the number cannot be stored in the numeric type supplied because of an overflow or other loss of precision.

### Boolean

Booleans can be automatically converted to Go's `bool` type (or any aliases of
it, like `type MyBoolean bool`) as long as the boolean value is not null or
unknown.

### List

Lists can be automatically converted to any Go slice type (or alias of a Go
slice type, like `type MyList []string`), with the elements either being
`attr.Value` implementations or being converted according to these rules. Go
slice types are considered capable of handling null values; the slice will be

> set to nil. Unknown list values will still cause an error to be returned.

Reviewer 2: ```suggestion
set to nil. `GET` will still return an error for unknown list values.
```

### Map

Maps can be automatically converted to any Go map type with string keys (or any
alias of a Go map type with string keys, like `type MyMap map[string]int`),
with the elements either being `attr.Value` implementations or being converted
according to these rules. Go map types are considered capable of handling null

> values; the map will be set to nil. Unknown map values will still cause an
> error to be returned.

Reviewer 2: ```suggestion
values; the map will be set to nil. `GET` will still return an error for unknown map values.
```
